src/kernel/capinv.c

```diff
@@ -434,8 +433,12 @@ cap_switch_thd(struct pt_regs *regs, struct thread *curr, struct thread *next,
 		return 0;
 	}
 
-	copy_gp_regs(regs, &curr->regs);
-	__userregs_set(&curr->regs, 0, __userregs_getsp(&curr->regs), __userregs_getip(&curr->regs));
+	if (!(curr->state & THD_STATE_PREEMPTED)) {
```

Comment:
We talked about modifying the return path in the `entry.S` file.  I'll need to think through the implications of this change.  The downside is that it slows down the thread switch path.